### DISCUSSION TITLE
API returns incomplete results with specific repositories

### DISCUSSION THREAD

**Discussion by carolkgn** [community/community#60518] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Hello!

I am using the REST API, with an organization-level token to extract data from the private repositories we have inside the organization.

I am using Python SDK and the requests library to extract this data, so my problem is: whether I use a sleep statement or not to request the data from the API, some of the repositories are returned with incomplete data.

Three different situations are happening:

1º. The repository returns without URL,  description, topics, and other personalized columns (that I think are out of scope in this question), but shows id_user, user login, readme title, readme size, and repository name;
2º. The repository doesn't return the user login and id_user, but shows the other information;
3º. The repository returns just two or three columns (in general, the repository name and the extraction date, columns that aren't extracted directly from the API), and nothing more.

What could be happening? This can be due to the use of different methods to extract information? Or possibly is a problem with secondary limits?

I already tried using a sleep statement to request from the API, but it didn't solve the problem.

Important information: I observed it is always the same repositories that have this type of problem, and eventually the list of problematic repos gets bigger with the creation of new repositories.

I appreciate any new insight, as everything I have tried and read on the documentation and here on the community page, hasn't worked until now.

**Comment by drahamim** [comment]
What Endpoints are you trying to use? Which Python SDK are you using?

**Reply by carolkgn** [reply]
I am trying to extract three main informations, the general list of repositories from the organization, with the details about these repositories that the API provides (description, topics and name in special), the list of contributors of each repository, the readme information about each repository, mainly the presence of a README file and the size of the README file.

The Python SDK I am using is [**gh api** ](https://cli.github.com/manual/gh_api), and some parts of the code use the library **requests** for Python and the cURL pattern of extraction used with the REST API.

**Reply by drahamim** [reply]
The gh api you linked is a command line tool. I would suggest using an actual python library such as 
[fastai/ghapi](https://github.com/fastai/ghapi)
or [pygithub](https://github.com/PyGithub/PyGithub). 
Both of these are based around the REST Github API. 

I haven't tried this one out yet but it supports both Rest and GraphQL APIs in one library. 
[githubkit](https://github.com/yanyongyu/githubkit)

These are the best option for pre-built that I have used in the past. 

Alternatively build it yourself using just the requests library itself directly.

**Reply by carolkgn** [reply]
I have inserted the wrong link in my previous message, I am using the [https://ghapi.fast.ai/](https://ghapi.fast.ai/) and requests library for Python.

I extract different information from both. I know how to extract it, and I am already extracting information with my code. But I am facing the problems I described, that's  why I am looking for insights of what's is going on.

**Reply by drahamim** [reply]
Can you post some of your code for the first issue?
I can better help from there.

**Reply by drahamim** [reply, marked as answer]
My code was specifically getting just the names for the repos but here it is. 
When i run this using the ghapi CLI ability i get quite a bit of info about a given repository from an Org level. 

```python
from ghapi.all import GhApi
from operator import itemgetter
api = GhApi()
......

 def get_org_repos(organization):
        """Call Github API and get list of repos
        for a given organization.

        Args:
            organization (string): name of github organization

        Returns:
            list: list of names of repositories in given organization
        """
        api_res = api.repos.list_for_org(org=organization, per_page=100)
        repo_list = list(map(itemgetter("name"), api_res))
        return repo_list
        
```

For your 2nd issue I think you are expecting the same keys in the get repo. I don't think user_id is a key in the json returned from `repos.get()` 

One of the things that is a bit weird with the API is that there is some switching between user/owner. and login/name 
I'm sure there is a logic to it but I don't fully understand when one is used over the other.

**Reply by carolkgn** [reply]
Hi! Thanks for your answer! I will try your solution! I also think that is some kind of logic I'm might missing about the repositories like the owner /user one.

So the code I'm using to extract the basic information from the API is this:

```
class KitHub:
    """
    Access Github API and extracts informations from existing repositories.
    """

    def __init__(self, key: str):
        self.key = key
        self.org = 'org'
        self.BaseUrl = 'https://api.github.com/orgs'
        self.headers = {'Authorization': f'token {key}',
                        'Accept': 'application/vnd.github.v3+json'}

    def GetRepos(self):
        """Gets the list of repositories.
        """

        PartUrl = 'repos'
        endpoint = os.path.join(self.BaseUrl, self.org, PartUrl)
        r = requests.get(endpoint, headers=self.headers)
        
        result_list = json.loads(r.content)
        status = r.status_code

        while 'next' in r.links.keys():
           
            endpoint_page = r.links['next']['url']
            
            r = requests.get(endpoint_page, headers=self.headers)
            time.sleep(5)
            page_result_list = json.loads(r.content)

            result_list += page_result_list

        self.GetReposResult = result_list
        self.GetReposStatus = status

        return
```
This is used on a jupyter notebook as an imported module. 

The code of the other problem, with the contributors, it's just a call to Gh API, like this:

```
github_key= 'token'

api = GhApi(owner='org', token=github_key) 

repo_name  = pd.DataFrame(api.repos.list_contributors(repo='repo_name'))
repo_name['repo'] = 'repo_name'
```
So the code is not pretty different from a basic API call in Python, but the pieces of information it returns to me are sometimes incomplete now.

**Reply by drahamim** [reply]
the first one looks fine to get information on an org. (It appears you are accessing [Get an Organization endpoint](https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#get-an-organization) 

The second one I've always done it with the org and repo in the repos.list_contributors() call rather then the Initialization of the GhApi. 
Maybe that would help? the GhAPI documentation wants this 
`repos.list_contributors(owner, repo, anon, per_page, page)`
to give you this output https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-contributors

**Reply by carolkgn** [reply]
Hello!

Thanks for your help!  

I tried using `api.repos.list_for_org(org=organization, per_page=100)`  and it showed information I was not receiving with the requests library approach, so thank you!

The problems with the list of contributors still persist, I guess the reason is not the code, but something else,  so I'll keep looking for the solution.
